osu.Game/Rulesets/UI/RulesetContainer.cs

```diff
@@ -43,7 +42,7 @@ public abstract class RulesetContainer : Container
         /// <summary>
         /// The input manager for this RulesetContainer.
         /// </summary>
-        internal readonly PlayerInputManager InputManager = new PlayerInputManager();
+        internal IHasReplayHandler ReplayInputManager => (IHasReplayHandler)KeyBindingInputManager;

```

Comment:
The _current_ rulesets guarantee this type. Someone making a ruleset would experience this invalidcast. If this was guaranteed then KeyBindingInputManager shouldn't be a PassthroughInputManager, but rather a default implementation of RulesetInputManager that does nothing.